climada/engine/uncertainty.py

```diff
@@ -0,0 +1,592 @@
+
+        Returns
+        -------
+        int
+            effective number of runs
+
+        """
+
+        if isinstance(self.params, pd.DataFrame):
+            return self.params.shape[0]
+        else:
+            return 0
+    @property
+    def param_labels(self):
+        """
+        Labels of all (exposure, impact function, hazard) uncertainty
+        parameters.
+
+        Returns
+        -------
+        list of strings
+            Labels of all uncertainty parameters.
+
+        """
+        return self.exp.labels + self.haz.labels + self.impf.labels
+
+    @property
+    def distr_dict(self):
+        """
+        Dictionnary of all (exposure, imapct function, hazard) distributions.
+
+        Returns
+        -------
+        distr_dict : dict( sp.stats objects )
+            Dictionnary of all distributions.
+
+        """
+
+        distr_dict = dict(self.exp.distr_dict)
+        distr_dict.update(self.haz.distr_dict)
+        distr_dict.update(self.impf.distr_dict)
+        return distr_dict
+
+    def make_sobol_sample(self, N, calc_second_order):
+        """
+        Make a sobol sample for all parameters with their respective
+        distributions.
+
+        Parameters
+        ----------
+        N : int
+            Number of samples as defined in SALib.sample.saltelli.sample().
+        calc_second_order : boolean
+            if True, calculate second-order sensitivities.
+
+        Returns
+        -------
+        None.
+
+        """
+
+        self.n_samples = N
+        sobol_uniform_sample = self._make_uniform_sobol_sample(
+            calc_second_order=calc_second_order)
+        df_params = pd.DataFrame(sobol_uniform_sample, columns=self.param_labels)
+        for param in list(df_params):
+            df_params[param] = df_params[param].apply(
+                self.distr_dict[param].ppf
+                )
+        self.params = df_params
+
+
+
+    def _make_uniform_sobol_sample(self, calc_second_order):
+        """
+        Make a uniform sobol sample for the defined model uncertainty parameters
+        (self.param_labels)
+        https://salib.readthedocs.io/en/latest/api.html#sobol-sensitivity-analysis
+
+        Parameters
+        ----------
+        calc_second_order : boolean
+            if True, calculate second-order sensitivities.
+
+        Returns
+        -------
+        sobol_params : np.matrix
+            Returns a NumPy matrix containing the sampled uncertainty parameters using
+            Saltelli’s sampling scheme.
+
+        """
+        problem = {
+            'num_vars' : len(self.param_labels),
+            'names' : self.param_labels,
+            'bounds' : [[0, 1]]*len(self.param_labels)
+            }
+        self.problem = problem
+        sobol_params = saltelli.sample(problem, self.n_samples,
+                                       calc_second_order=calc_second_order)
+        return sobol_params
+
+
+    def calc_impact_sobol_sensitivity(self,
```

Comment:
Why not a single kwarg, e.g. ooi (outcomes of interest) like ooi={calc_eai_exp=bool, calc_at_event=bool, rp=list} where only those of interest are listed (with all possible outcomes being all those supported by Impact().calc, which the user should be aware of) and then we assign e.g. self.calc_eai_exp or self.calc_at_event based on those we have. These self.... will also dictate what calculations we save (as it is done already).

Reply:
Good idea. I will see how it works out.